---
title: "Introduction to the pedtools package"
author: "Magnus Dehli Vigeland"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette:
    toc: true
    toc_depth: 2
    number_sections: true
vignette: >
  %\VignetteIndexEntry{Introduction to the pedtools package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

The purpose of this vignette is to show how to work with pedigrees and marker data in `pedtools`. To obtain the latest version of the package I recommend installing it from GitHub as follows:
```{r, eval = FALSE}
 # First install devtools if needed
if(!require(devtools)) install.packages("devtools")

# Install pedtools from github
devtools::install_github("magnusdv/pedtools")
```

Now you should be able to load `pedtools`:
```{r message=FALSE}
library(pedtools)
```


# Pedigrees
To create a pedigree in pedtools, the most direct method is to use the `ped()` function. The essential arguments to `ped()` are 4 vectors of equal length:

* `id` : individual ID labels (numeric or character)
* `fid` : id of the fathers (0 if not included)
* `mid` : id of the mothers (0 if not included)
* `sex` : gender code, either 0 (unkown), 1 (male) or 2 (female)

Hence, `id[j]` denotes the ID of the j'th pedigree member, whose parents are `fid[j]` and `mid[j]`, and whose gender is given by `sex[j]`.

The following creates a family trio, i.e. father, mother and child.

```{r}
ped(id = 1:3, fid = c(0,0,1), mid = c(0,0,2), sex = c(1,2,2))
```

The child (id=3) is female, since the associated entry in `sex` is 2. Note that missing parents are printed as `*`. A member without parents is called a _founder_ of the pedigree, while a _nonfounder_ has both parents specified. It is not allowed to have exactly one parent. 

Instead of using numerical ID labels as above, we could have used characters. Let us create the trio again with more informative labels, and store it in a variable named `trio`.
```{r}
trio = ped(id = c("fa", "mo", "girl"), fid = c(0,0,"fa"), mid = c(0,0,"mo"), sex = c(1,2,2))
trio
```

To plot the result, simply use `plot`:
```{r}
plot(trio)
```

From the way it is printed, the object `trio` appears to be a data frame, but this is not exactly true. Rather it is an object of class `ped`, which is basically a list. We can see the actual content of `trio` by unclassing it:
```{r}
unclass(trio)
```

## Built-in shortcuts for constructing pedigrees
Rather than using the `ped()` function directly, it is usually quicker and safer to build pedigrees step by step, applying the arsenal of utility functions offered by `pedtools`. A typical workflow is as follows:

1. Choose one of the basic pedigree structures as starting point.
1. Add and/or remove individuals as needed.
1. Modify attributes like genders and labels.

Before showing some examples, we list the available tools for each of the 3 steps. Most of these have a very specific purpose and are designed with user-friendlyness in mind. After reading the examples below you should be ready to start building your own pedigrees. Consult the help pages when you get stuck.

**Basic pedigree structures**  
The following pedigree structures serve as starting points for pedigree constructions. For parameters and details, see `?ped_basic`.

* `singleton()`, creates a pedigree consisting of a single individual.
* `nuclearPed()`, creates a nuclear pedigree (parents+children)
* `cousinsPed()`, a pedigree with cousins of specified degree/removal
* `halfCousinsPed()`, a pedigree with half cousins of specified degree/removal

There are also more specialized structures, including double cousins and breeding schemes like consequtive sib mating. Look them up in `?ped_complex` if you are interested.

**Add/remove individuals**  
The functions below are used to add or remove members of a pedigree. For details, see `?ped_add`.

* `addChildren()`
* `addSon()` and `addDaughter()` (special cases of addChildren())
* `addParents()`, 
* `removeIndividuals()`
* `branch()`

**Modify attributes**  
The following functions modify various attributes of a `ped` object. See `?ped_modify` for parameters and details. 

* `swapSex()`
* `relabel()`
* `setLabels()` (deprecated)
* `setFamid()` 

## Examples of pedigree construction
Let us recreate the `trio` pedigree without using `ped()`. The obvious starting point for this is `nuclearPed()`, and we set `nch=1` to indicate 1 child. By default, this creates a trio with numeric labels (father=1; mother=2; child=3) and a male child. Hence we fix the gender with `swapSex()`, and edit the labels with `relabel()`:

```{r}
trio2 = nuclearPed(nch = 1)
trio2 = swapSex(trio2, id = 3)
trio2 = relabel(trio2, new = c("fa", "mo", "girl"))
```
In fact, all of this may be done in a single call to `nuclearPed()`. Both child genders and labels can be specified directly:
```{r}
trio3 = nuclearPed(father = "fa", mother = "mo", children = "girl", sex = 2)
```
We can confirm that both results are identical to the original `trio`:
```{r}
identical(trio, trio2)
identical(trio, trio3)
```

Of course, there are many other ways of building this simple pedigree. Here is another:
```{r}
trio4 = singleton("fa")
trio4 = addDaughter(trio4, parent = "fa", id = "girl")
trio4 = relabel(trio4, old = "NN_1", new = "mo")
```

Note that `addDaughter()` created the mother as "NN_1", so we needed to relabel her.  
(A digression about  `relabel()`: If the parameter `old` is missing, as when we made `trio2` above, it defaults to all members, in the order you see when printing the pedigree.)

## Pedigree subsets
Pedtools offers a range of utility functions for identifying subsets of pedigree members. These come in two flavours: 1) members with certain _global_ property, and 2) members with a certain relationship to a given individual.

**Pedigree members with a certain property**  
Each of the following functions returns a vector specifying the members with the given property. 

* `founders()`
* `nonfounders()`
* `leaves()`
* `males()`
* `females()`
* `typedMembers()`
* `untypedMembers()`

By default, the output of these functions is a character vector containing ID labels. However, adding the option `internal = TRUE` will give you an integer vector instead, reporting the internal indices of the members. This is frequenctly used in the source code of `pedtools`, but is usually not intended for end users of the package.

**Relatives of a given individual**  
The functions below take as input a `ped` object and the label of a single member. They return a vector of all members with the given relation to that individual.

* `father()`
* `mother()`
* `parents()`
* `grandparents()`
* `children()`
* `spouses()`
* `siblings()`
* `cousins()`
* `nephews_nieces()`
* `ancestors()`
* `descendants()`
* `unrelated()`

# Markers

The other main theme of the `pedtools` package (_pedigrees_ being the first) are marker genotypes. 

## Creating marker objects
Marker objects created with the `marker()` function. For example, the following command makes an empty marker  associated with the `trio` pedigree:
```{r}
marker(trio)
```
As shown in the output, the marker is indeed empty: All pedigree members have missing genotypes, and there is no assigned name or chromosome/position. Furthermore, the last lines show that there is only one allele (named "1"), with frequency 1. For a more interesting example, let us make a SNP named "snp1", with alleles "A" and "B". The father is homozygous "A/A", while the mother is heterozygous. We store it in a variable `m1` for later use.
```{r}
m1 = marker(trio, fa = "A", mo = c("A","B"), name = "snp1")
```
This illustrates several points. Firstly, individual genotypes are specified using the ID labels. For homozygous genotypes it suffices to write the allele once. Furthermore, the different alleles occuring in the genotypes is interpreted as the complete set of alleles for the marker. Finally, these are assigned equal frequencies. Of course, this behavious can be overridden, by declaring alleles and frequencies explicitly:
```{r}
marker(trio, fa = "A", mo = c("A","B"), alleles = c("A","B","C"), afreq = c(.2,.3,.5))
```
The markers chromosome can be declared using the `chrom` argument, and similarly its position by `posMb` (megabases) and/or `posCm` (centiMorgan). Markers with unknown chromosome are treated as autosomal. To define an X-linked marker, put `chrom=23`. the fact that males are hemizygous on X (i.e. they have only one allele) is reflected in the printout of such markers:

```{r}
m2 = marker(trio, fa = "A", mo = c("A","B"), chrom = 23, name = "snpX")
```

A side note: It may come as a surprise that you don't need quotes around the ID labels (which are characters!) in the above commands. This is because `marker()` uses _non-standard evaluation (NSE)_, a peculiarity of the R language which often leads to less typing and more readable code.[^1] Unfortunately, this doesn't work with numerical ID labels. Thus to assign a genotype to someone labelled "1" you need quotes, as in `marker(trio, "1"="A")`.

[^1]: You may have come across NSE before, for instance when using `subset()` on a data.frame. To learn more about NSE, I highly recommend this book chapter by Hadley Wickham:  
<http://adv-r.had.co.nz/Computing-on-the-language.html>


## Plotting pedigrees with marker data
Including marker data in a pedigree plot is straightforward:
```{r}
plot(trio, marker = m1)
```

## Markers attached to ped objects
Although a `ped` object is needed in the creation of a `marker`, the two are independent of each other once the marker is created. In many applications it is useful to _attach_ markers to their `ped` object. In particular for bigger projects with many markers, this makes it easier to manipulate the dataset as a unit. 

To attach a marker `m` (which could be a list of several markers) to a pedigree `x`, there are two options:

* `setMarkers(x, m)`
* `addMarkers(x, m)`

The difference between these is that `setMarkers()` replaces all existing markers, while `addMarkers()` appends `m` to the existing ones. In our `trio` example the two are equivalent since there are no existing markers.
```{r}
trio = setMarkers(trio, list(m1, m2))
trio
```

**Selecting and removing attached markers**  
Four closely related functions functions are useful for manipulating markers attached to a `ped` object:

* `selectMarkers()` : returns a `ped` object where only the indicated markers are retained
* `removeMarkers()` : returns a `ped` object where the indicated markers are removed
* `getMarkers()` : returns a list of the indicated markers
* `whichMarkers()` : returns the indices of the indicated markers

All of these have exactly the same arguments, described in more detail in `?marker_select`. Let us do a couple of examples here. Recall that by now, our `trio` has two attached markers.
```{r}
whichMarkers(trio, chrom = 23)
selectMarkers(trio, markers = "snp1")
```

## Accessing and modifying marker attributes
Internally, a marker object is stored as a matrix with two columns (one for each allele) and one row for each pedigree member. The matrix is numeric (for computational convenience) while the allele labels and other meta information are added as _attributes_. The most important of these are:

* `alleles` : The allele labels, stored as a character vector.
* `afreq` : The allele frequencies, in the same order as the alleles. An error is issued if the frequencies do not sum to 1 after rounding to 3 decimals.
* `name` : The marker name, which can be any character string not consisting solely of digits.
* `chrom` : The chromosome name. This can be given as an integer, but is always converted to character. The special values "23" and "X" are recognized as the human X chromosome, which affects the way genotypes are printed.
* `posMb` : Chromosomal position given in megabases.
* `posCm` : Chromosomal position given in centiMoran.

In addition to those listed above, there are two more attributes: `pedmembers` and `sex`. They store the ID labels and genders of the pedigree associated with the marker, and are only used to empower the printing method of marker objects.

**Marker accessor functions**  
For each marker attribute listed above, there is a corresponding function with the same name for retrieving its content. These functions take as input either a `marker` object, or a `ped` object together with the name (or index) of an attached marker. This may sound a bit confusing, but a few examples will make it clear! 

Recall that our marker "snp1" is stored in the variable `m1`, but also attached to `trio`. To extract its allele frequencies we therefore have two options, both calling the function `afreq()`:

```{r}
afreq(m1)
afreq(trio, marker = "snp1")
```

We can also _modify_ the frequencies using this syntax. To avoid confusion about the order, the frequencies must be named with the allele labels (just as in the output of `afreq()` above). 
```{r}
afreq(trio, marker = "snp1") = c(A = 0.9, B = 0.1)
```

In addition to the functions getting and setting marker attributes, there is one marker accessor, namely `genotype`. This returns the genotype of a specified individual, and can also be used to modify genotypes. As the others, it can be applied to marker objects directly, or to pedigrees with attached markers. Here we show a few examples of the latter type:
```{r}
genotype(trio, "snpX", id = "girl")
genotype(trio, "snpX", id = "girl") = "A"
trio

```

